docs/DeveloperGuide.md

```diff
@@ -296,6 +296,151 @@ The following sequence diagram shows how the flow of the example execution:
+
+Realising the commonality between all kinds of edit commands, we have created an abstract generic class, `EditCommand`, to serve as template class containing the logic of editing a `ListEntry` object.
+
+The signature of the `EditCommand` class is as follows:
+```
+public abstract class AbstractEditCommand<T extends ListEntry<? extends T>> extends Command
+```
+And the concrete edit classes (`EditPersonCommand` class for example) can be simply implemented by "filling in the blank" and providing the correct parameters to the template class.
+
+The template for the edit process is as follows:
+```
+initModelMethods();
+init();
+editFields();
+updatePersonLessonMap();
+validateEditedAndWriteBack();
+showMethod.accept(edited);
+return new CommandResult("Edit success.\n from: " + original.toString() + "\n to: " + edited.toString());
+```
+The `EditPersonCommand` class can be simply implemented by overriding the abstract methods in the `EditCommand` class.
+
+The most important method to be overridden is the initModelMethods(), which looks like this in the `EditPersonCommand` class:
+```
+ @Override
+    protected void initModelMethods() {
+        currentShownEntry = model.getCurrentlyDisplayedPerson();
+        list = model.getFilteredPersonList();
+        hasClashWith = model::hasPersonClashWith;
+        deleteMethod = model::deletePersonForEdit;
+        addMethod = model::addPerson;
+        getClashingEntry = model::getPersonClashWith;
+        showMethod = model::showPerson;
+    }
+```
+
+#### Design considerations:
+
+**Aspect: How to keep the code clean and maintainable while providing the flexibility and variety of valid user input while .
+
+* **Alternative 1 (current choice):** Group the parsing and execution of a command into four groups of classes: `AddressBookParser`, `TypeParsingUtil`, `EditPersonCommandParser` and `EditPersonCommand`.
+    * Pros:
+        * Clear separation of concerns.
+        * Easy to maintain, easy to extend as for each new `ListEntryField` and `ListEntry` class that is added and fulfill the contract of the `ListEntryField` and `ListEntry` class, the parsing and execution of the command will almost be automatically supported.
+        * Good DRY principle. Only one method is needed to parse all kinds of flags, the knowledge of a given type of processing is encapsulated in a single place.
+    * Cons: Less flexible. It is not easy to add flags that has special parsing logic not supported by the current template.
+
+* **Alternative 2:** Having even stricter constraints on the `ListEntryField` and `ListEntry` class, so that we can use reflection to automatically parse the flags.
+    * Pros: Potentially better extensibility.
+    * Cons:
+        * Difficult to implement.
+        * Even more difficult to debug.
+        * Even less flexible.
+        * Not easy to get the correct error message, as reelection call will through a `InvocationTargetException` instead of a `ParseException` when the parsing fails.
```

Comment:
what is a reelection call
i think through is supposed to be throw?

Reply:
Thank you for pointing that out. Rephrased in the latest PR